src/websocket/stream.py

```diff
@@ -121,86 +132,104 @@ async def echo(self, websocket: WebSocket) -> None:
-                result = {"result": words, "text": " ".join([x["word"] for x in words])}
-                self.last_final_result_object = result
-                self.last_final_bytes = chunk_cache
-                self.final_transcriptions.append(result)
-
-            if not skip_send:
+            agreed_results = []
+            if self.agreement.contains_has_sentence_end():
+                agreed_results = self.agreement.flush_at_sentence_end()
+            else:
+                agreed_results = self.agreement.flush_confirmed()
+
+            result = await self.build_result_from_words(agreed_results)
+            # The final did not contain anything to send
+            if len(result['result']) == 0:
+                return
+            self.final_transcriptions.append(result)
+
+            # Shorten window if needed
+            if len(self.sliding_window) > MAX_WINDOW_SIZE_BYTES:
+                bytes_to_cut_off = len(self.sliding_window) - MAX_WINDOW_SIZE_BYTES
+                self.logger.debug(f"Reducing sliding window size by {bytes_to_cut_off} bytes")
+                self.previous_byte_count += bytes_to_cut_off
+                self.window_start_timestamp += bytes_to_cut_off / BYTES_PER_SECOND
+                self.sliding_window = self.sliding_window[bytes_to_cut_off:]
+
+            if not self.close_stream:
                 await websocket.send_text(json.dumps(result, indent=2))
+
             self.logger.debug(
-                f"Final Transcription took {time.time() - start_time:.2f} s"
+                f"Published final of {len(agreed_results)}."
             )
+            self.last_final_published = time.time()
 
         except Exception:
             self.logger.error(
                 f"Error while transcribing audio: {traceback.format_exc()}"
             )
             self.close_stream = True
 
-    async def transcribe_chunk_partial(
-        self, websocket, chunk_cache, recent_cache
+    async def build_result_from_words(self, words: List[Word],save=True) -> Dict:
+        overall_transcribed_seconds = self.previous_byte_count / BYTES_PER_SECOND
+
+        cutoff_timestamp = 0
+        if len(self.final_transcriptions) > 0:
+            cutoff_timestamp = self.final_transcriptions[-1]["result"][-1]["end"]
+
+        result = {"result": [], "text": ""}
+        for word in words:
+            start = float(f"{word.start:.6f}") + overall_transcribed_seconds
+            end = float(f"{word.end:.6f}")  + overall_transcribed_seconds
+            conf = float(f"{word.probability:.6f}")
+            if end <= cutoff_timestamp + 0.01 and save:
+                continue
+            result["result"].append(
+                {
+                    "conf": conf,
+                    # the start time and end time is the time of the word minus the time of the current final
+                    "start": start,
+                    "end": end,
+                    "word": word.word.strip(),
+                }
+            )
```

Comment:
superior list comprehension

if not applied, make comment first letter uppercase please <3


```suggestion
        result["result"] = [
    {
        "conf": float(f"{word.probability:.6f}"),
        "start": float(f"{word.start:.6f}") + overall_transcribed_seconds,
        "end": float(f"{word.end:.6f}") + overall_transcribed_seconds,
        "word": word.word.strip(),
    }
    for word in words
    if float(f"{word.end:.6f}") > cutoff_timestamp + 0.01
]
```